lib/Onyx.js

```diff
@@ -777,19 +777,75 @@ function addKeyToRecentlyAccessedIfNeeded(mapping) {
  * @param {Object} mapping
  */
 function getCollectionDataAndSendAsObject(matchingKeys, mapping) {
-    Promise.all(_.map(matchingKeys, (key) => get(key)))
-        .then((values) =>
-            _.reduce(
-                values,
-                (finalObject, value, i) => {
-                    // eslint-disable-next-line no-param-reassign
-                    finalObject[matchingKeys[i]] = value;
-                    return finalObject;
-                },
-                {},
-            ),
-        )
-        .then((val) => sendDataToConnection(mapping, val, undefined, true));
+    // Keys that are not in the cache
+    const missingKeys = [];
+    // Tasks that are pending
+    const pendingTasks = [];
+    // Keys for the tasks that are pending
+    const pendingKeys = [];
+
+    // We are going to combine all the data from the matching keys into a single object
+    const data = {};
+
+    /**
+     * We are going to iterate over all the matching keys and check if we have the data in the cache.
+     * If we do then we add it to the data object. If we do not then we check if there is a pending task
+     * for the key. If there is then we add the promise to the pendingTasks array and the key to the pendingKeys
+     * array. If there is no pending task then we add the key to the missingKeys array.
+     *
+     * These missingKeys will be later to use to multiGet the data from the storage.
+     */
+    matchingKeys.forEach((key) => {
+        const cacheValue = cache.getValue(key);
+        if (cacheValue) {
+            data[key] = cacheValue;
+            return;
+        }
+
+        const pendingKey = `get:${key}`;
+        if (cache.hasPendingTask(pendingKey)) {
```

Comment:
We can easily do one cache lookup instead of two here:

```js
const pendingKey = `get:${key}`;
const pendingTask = cache.getPendingTask(pendingKey);
if (pendingTask) {
    pendingTasks.push(pendingTask)
}
```